### PR TITLE
Connectionist temporal classification (CTC)

### PR DESCRIPTION
**Pull request by cifkao**
- Graves, Alex, et al., 2006. ["Connectionist temporal classification: labelling unsegmented sequence data with recurrent neural networks."](http://www.cs.toronto.edu/~graves/icml_2006.pdf)
- `PlainRunner`: Relies on `decoder.decoded` instead of actively decoding the output.
- Test using the [Yesno](http://www.openslr.org/1/) speech dataset from OpenSLR. (Maybe it can be included in the repo?)

### CODE REVIEW

**Comment by jindrahelcl** [comment]
Hustý.. Akorát se mi moc nelíbí ten název toho plain runneru. Pro ``encoded``od enkodéru máme něco, co se jmenuje ``RepresentationRunner``, což se mi taky moc nelíbí, ale přinejmenším bych tyhle dva názvy nějak zkonzistentnil.. Nešlo by to spíš udělat tak, že každá ``ModelPart`` bude mít něco jako výstupní tenzor (`encoded` pro enkodéry a `decoded` pro dekodéry) a udělal by se společnej runner na tyhle výstupní tenzory?

**Comment by cifkao** [comment]
`encoded` je teď poslední stav enkodéru, ne? Výstupy enkodéru bývají v `hidden_states`, resp. `_attention_tensor` (tam s dropoutem?) a jejich maska v `_attention_mask`, resp. já jsem ji teď ještě pojmenoval `states_mask`, abych k ní mohl z dekodéru... Uznávám, že by se to mělo nějak sjednotit.

**Comment by jlibovicky** [comment]
Co je zač ten _yes/no_ dataset? Jestli má nějakou schůdnou licenci, tak bych to možná přidal do `tests/data`.

**Comment by cifkao** [comment]
@jlibovicky "No formal license but free to use for any purpose."

Rozbalené to má 5.8 MB.

**Comment by jindrahelcl** [comment]
Tak přidej do testů úryvek z toho celýho datasetu, třeba 50 examplů, a to se tam dá commitnout. 5megový soubory bych sem nedával.

**Comment by cifkao** [comment]
@jindrahelcl @jlibovicky Ono už takhle je to jen 60 příkladů, ale holt jsou to WAVy.

**Comment by cifkao** [comment]
http://www.openslr.org/1/

**Comment by cifkao** [comment]
Dal jsem jich sem 15, každý má kolem 100 KB.

File: neuralmonkey/encoders/sentence_encoder.py [hunk 1]
```diff
@@ -117,6 +117,10 @@ def _attention_mask(self):
         return self.input_mask
 
     @property
+    def states_mask(self):
+        return self.input_mask
+
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_encoder.py]: liší se někdy states mask a input mask? (Nemyslim v sentence enkodéru)

Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_encoder.py]: Jestli ne, dal bych enkodérům (respektive všem `ModelPart` s nějakým sekvenčním vstupem) `input_mask` a všechno maskoval podle toho. `attention_mask` je použitá na pár místech a přejmenovat to je jednoduchý.

**Comment by cifkao** [comment]
Liší se to v tom konvolučním enkodéru, kde se to zkrátí max poolingem.

**Comment by cifkao** [comment]
Liší se to v tom konvolučním enkodéru, kde se to zkrátí max poolingem.

On 11 Apr 2017 10:34, "Jindra Helcl" <notifications@github.com> wrote:

*@jindrahelcl* commented on this pull request.
------------------------------

In neuralmonkey/encoders/sentence_encoder.py
<https://github.com/ufal/neuralmonkey/pull/382#discussion_r110842351>:

> @@ -117,6 +117,10 @@ def _attention_mask(self):
         return self.input_mask

     @property
+    def states_mask(self):
+        return self.input_mask
+

liší se někdy states mask a input mask? (Nemyslim v sentence enkodéru)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/ufal/neuralmonkey/pull/382#pullrequestreview-32041916>,
or mute the thread

File: neuralmonkey/encoders/sentence_encoder.py (hunk 1 above)
Comment (jlibovicky) [inline neuralmonkey/encoders/sentence_encoder.py]: Já bych to tyhle rozdíly teď neřešil a někdy se sejdeme a rozmyslíme pořádnou hierarchii a dědičnost těch částí modelu.

Comment (jindrahelcl) [inline neuralmonkey/encoders/sentence_encoder.py]: ok
